orttraining/orttraining/core/framework/ortmodule_graph_builder.cc

```diff
@@ -78,15 +78,16 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
     SetConcreteInputShapes(*input_shapes_ptr);
   }
 
-  // Optimize the inference graph, and if needed, build the gradient graph.
-  std::unordered_set<std::string> x_node_arg_names;
-  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
+  // This graph will be used only for inferencing, so stop right here.
+  // No need to apply the optimizations for training or build a gradient graph.
   if (!config_.build_gradient_graph) {
-    // This graph will be used only for inferencing, so stop right here.
-    // No need to build a gradient graph.
-    return Status::OK();
+    // Save a copy of inference optimized model
+    return Model::Load(gradient_model_->ToProto(), inference_optimized_model_, nullptr, *logger_);
   }
 
+  // Optimize the inference graph and build the gradient graph.
+  std::unordered_set<std::string> x_node_arg_names;
+  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
```

Comment:
Inside the `OptimizeInferenceGraph`, there is a call to save the optimized model again inside `inference_optimized_model_`. Can we remove that call?

Also, may we rename the `OptimizeInferenceGraph` to `OptimizeForwardGraph`/`OptimizeEvalGraph` and `inference_optimized_model_` to `forward_optimized_model_`/`eval_optimized_model_`?

Reply:
I am not sure how the frontend uses such models_ in the class, so I would rather not the rename such model_ names and the saving logic. Changing OptimizeInferenceGraph is good suggestion to remove the ambiguity.

Reply:
to give some context: the ortmodule python code has two parts `TrainingManager` and `InferenceManager`. Both instantiate their own `OrtModuleGraphBuilder`s. `InferenceManager` sets the `OrtModuleGraphBuilderConfiguration.build_gradient_graph` to `False` while `TrainingManager` sets it to `True`. And then each of them has the onnx model corresponding to either training or eval.

Reply:
I went over the OrtModuleGraphBuilder and the frontend code and made a refactor. please check the new commit.